### DISCUSSION TITLE
Best practices for using pysr for binary classification?

### DISCUSSION THREAD

**Discussion by andrew-maris** [MilesCranmer/PySR#478] [Q&A] [answered]
Hi all,

I was wondering what are the best practices for using pysr for binary classification. I noticed there are some past discussions mentioning the topic, and references to classification losses in the documentation, but no complete guide for tackling the problem that I am aware of. What is the right way to use pysr in this case?

For some background, I'm studying a particular instability in magnetic fusion - the "density limit" - where the confinement of the plasma suddenly collapses when too much density is injected. There are some empirical & theoretical power law scalings for what conditions this occurs at, but I found in the database I assembled that they lead to poor classification accuracy (True positive rate of ~60% when false positive rate is 5%). I spent a good amount of time adding new features to the dataset that have not been used in other scalings, and I would like to use pysr to find a more reliable stability boundary (& compare the pysr model with non-interpretable models like a deep NN and interpretable models like a power law derived from a linear SVM).

In any case, I'm curious is the best way to formulate a classification problem using pysr. Has this been rigorously explored? I've played around with using a custom loss function, but I was curious if there's an accepted set of "best practices".

**Comment by MilesCranmer** [comment, marked as answer]
Hey @andrew-maris!

I think you might be able to do classification loss on your problem if you select one of these binary classification losses: https://astroautomata.com/SymbolicRegression.jl/stable/losses/#Classification, and use them for the `loss` argument (https://astroautomata.com/PySR/api/).

e.g., `loss="L2MarginLoss()"` which is the same as `loss="myloss(pred, target) = (1 - pred * target)^2"`. You assume here that `target` is either `+1` or `-1` (you would feed the data into `.fit(X, y)` where `y` is a bunch of `+1` and `-1`). So if you have `pred, target = (-1, -1)` or `pred, target = (1, 1)`, the loss will be minimized. 

Cheers,
Miles

**Reply by andrew-maris** [reply]
Thanks for the reply @MilesCranmer! I went ahead and tried it out, and found a reasonable-ish answer after a few thousand iterations.

A question that might help pysr explore the space faster: my data columns are all positive (for example, we have density and temperature), what would be a good way to help pysr to find a model with an output like that spans -1 to 1? It took a few thousand iterations for pysr to figure out an approximate way to do this on it's own: it found f(x) = T^{-0.3} - 2.2, which is a good one-variable fit to the data since the density limit event - the positive class - tends to occur when T<0.1. It took many iterations to get to that point, however, and I wonder if I could get a speed up by helping pysr naturally deal with this discrepancy between positive columns and negative labels.

I thought about subtracting the median value from each column, but there is no physical meaning to median and so I would like to avoid that. I also thought about customizing the loss function such that  the prediction is fed through a function that spans -1 and 1 (ex. predicted label = tanh[prediction - 1] ), but that seemed a little clunky. Of course, the simplest option is just to ignore this complication and increase the number of iterations. None of these feel particularly satisfactory, so I'm curious if you have thoughts on what usually works best!

**Reply by MilesCranmer** [reply]
For this kind of requirement I’d recommend writing a custom objective that imposes such constraints. Objectives do not need to be differentiable in PySR so you can really get advanced about it.

see https://astroautomata.com/PySR/examples/#9-custom-objectives for a custom evaluation scheme example.

see https://github.com/MilesCranmer/PySR/discussions/475#discussioncomment-7686039 for an example of limiting equations to exactly two sinusoids

I think you could do something similar here and perhaps try to impose specific behavior on the different features. (Maybe there’s a simpler strategy though, not sure).

**Reply by andrew-maris** [reply]
Great thanks, I'll give that a try.
